source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,184 @@
+module mir.combinatorics;
+
+import std.range.primitives: isRandomAccessRange, hasLength;
+import std.stdio;
+
+Permutations permutations(size_t n) @safe pure nothrow
+{
+    assert(n); //???
+    return Permutations(new size_t[n-1], new size_t[n]);
+}
+
+auto permutations(Range)(Range r) @safe pure nothrow
+if (isRandomAccessRange!Range && hasLength!Range)
+{
+    import std.range : indexed;
+    import std.algorithm: map;
+    size_t n = r.length;
+    assert(n); //???
+    return Permutations(new size_t[n-1], new size_t[n]).map!(p => r.indexed(p));
```

Comment:
The main problem is that can't use delegates in nogc and if we find a workaround it will probably difficult for the user too?

On 2016-04-11 09:16, Ilya Yaroshenko wrote:
>> +import std.stdio;
>> +
>> +Permutations permutations(size_t n) @safe pure nothrow
>> +{
>> +    assert(n); //???
>> +    return Permutations(new size_t[n-1], new size_t[n]);
>> +}
>> +
>> +auto permutations(Range)(Range r) @safe pure nothrow
>> +if (isRandomAccessRange!Range && hasLength!Range)
>> +{
>> +    import std.range : indexed;
>> +    import std.algorithm: map;
>> +    size_t n = r.length;
>> +    assert(n); //???
>> +    return Permutations(new size_t[n-1], new size_t[n]).map!(p => r.indexed(p));
> 
> I know, but Phobos is not the bible
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/DlangScience/mir/pull/65/files/aee1db667cadff4a4b20e2d59beed1c71e72c324#r59157830
>